src/cljc/ataru/virkailija/editor/form_diff.cljc

```diff
@@ -0,0 +1,223 @@
+(ns ataru.virkailija.editor.form-diff)
+
+(defn- index-of-id [id form]
+  (if-let [some-id id]
+    (if-let [some-match (ffirst (filter #(= id (:id (second %))) (map-indexed vector (:content form))))]
+      some-match
+      nil)))
+
+(defn- index-of-element [element form]
+  (let [id (:id element)]
+    (index-of-id id form)))
+
+(defn is-updated-in [old-element element]
+  (and old-element
+       (not (= old-element element))))
+
+(defn find-element [id form]
+  (first (filter #(= (:id %) id) (:content form))))
+
+(defn find-elements [ids form]
+  (filter #(contains? ids (:id %)) (:content form)))
+
+(defn find-updated-elements [old-form form]
+  (filter #(is-updated-in (find-element (:id %) old-form) %) (:content form)))
+
+(defn insert-adjacent-elements-between [adjacent-elements siblings form]
+  (let [content (:content form)
+        [above below] (doall (map #(index-of-id % form) siblings))]
+    (if (or
+         (and (not above) (first siblings))
+         (and (not below) (second siblings)))
+      (throw (ex-info "Sibling with index not found! Mismatching forms!" {})))
+    (cond
+     (and above
+          (not below)) (assoc form :content (concat content adjacent-elements))
+     (and (not above)
+          below) (assoc form :content (concat adjacent-elements content))
+     (and (not above)
+          (not below)) (assoc form :content adjacent-elements)
+     (and above
+          below) (let [adjacent? (= 1 (- above below))]
+                   (if adjacent?
+                     (let [[elements-above elements-below] (split-at above content)]
+                       (assoc form :content (concat elements-above adjacent-elements elements-below)))
+                     (throw (ex-info "Mismatching forms! User should update view." {})))))))
+
+(defn find-element-siblings [form element]
+  (let [content (:content form)
+        index (index-of-element element form)
+        sibling-above (get content (dec index))
+        sibling-below (get content (inc index))]
+    [sibling-above sibling-below]))
+
+(defn- element-with-new-siblings? [element old-form form]
+  (let [old-siblings (find-element-siblings old-form element)
+        new-siblings (find-element-siblings form element)]
+    (not (= (map :id old-siblings) (map :id new-siblings)))))
+
+(defn find-elements-with-new-siblings
+  [old-form form]
+  (filter #(element-with-new-siblings? % old-form form) (:content form)))
+
+(defn find-missing-elements
+  [old-form form]
+  (let [new-ids (set (map :id (:content form)))
+        old-ids (set (map :id (:content old-form)))
+        keys-in-old (clojure.set/difference old-ids new-ids)]
+    (filter #(contains? keys-in-old (:id %)) (:content old-form))))
```

Comment:
Tää `filter` on tarpeeton, `keys-in-old` on jo juurikin ne id:t jotka on `old-form`:ssa muttei `form`:ssa.